src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperContext.cs

```diff
@@ -25,12 +26,13 @@ public class TagHelperContext
         /// <param name="getChildContentAsync">A delegate used to execute and retrieve the rendered child content 
         /// asynchronously.</param>
         public TagHelperContext(
-            [NotNull] IDictionary<string, object> allAttributes,
+            [NotNull] IEnumerable<IReadOnlyTagHelperAttribute> allAttributes,
             [NotNull] IDictionary<object, object> items,
             [NotNull] string uniqueId,
             [NotNull] Func<Task<TagHelperContent>> getChildContentAsync)
         {
-            AllAttributes = allAttributes;
+            AllAttributes = new ReadOnlyTagHelperAttributes<IReadOnlyTagHelperAttribute>(
+                allAttributes.Select(attribute => new TagHelperAttribute(attribute.Name, attribute.Value)));
```

Comment:
Can't do it there, the `ReadOnlyTagHelperAttributes` is generic with the constraint that the `TAttribute` is of type `IReadOnlyTagHelperAttribute`. Therefore you can't do:

```C#
public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttribute> attributes)
{
    Attributes = new List<TAttribute>(
        attributes.Select(attribute => new TAttribute
        {
            Name = attribute.Name,
            Value = attribute.Value
        }));
}
```

Even if you add the `new()` constraint.